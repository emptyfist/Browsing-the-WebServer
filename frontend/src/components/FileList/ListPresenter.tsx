import React, {useState} from 'react';
import toast from 'react-hot-toast';
import { FileInfo } from '../../types';
import { useLocation, useNavigate } from 'react-router-dom';
import { fetchWrapper } from '../../helpers/fetch-wrapper';
import WaitingComponent from '../Waiting/Waiting';
import ListEntry from './ListEntry';
import ConfirmModal from '../ConfirmModal';
import { ListResponse } from '../../types';

interface Props {
  isLoading: boolean,
  list: FileInfo[],
  refresh: (path: string) => void,
};

const ListPresenter = ({isLoading, list, refresh}: Props) => {
  const location = useLocation();
  const navigate = useNavigate();
  const [selName, setSelName] = useState('');

  const showConfirmDlg = (name: string) => {
    setSelName(name);
  }

  const closeModal = async (isDelete: boolean) => {
    if (isDelete) {
      const result = await fetchWrapper.delEntry(`/api/file/delete?path=${location.pathname === '/' ? '' : location.pathname}/${selName}`)
        .then((res:ListResponse) => {
          if (res.Result === false) {
            toast.error(`Error message from Server : ${res.ErrorMessage}`);
          } else {
            toast.success(`${selName} is deleted`);
            return true;
          }
        }).catch((err: ListResponse | undefined) => {
          toast.error("Error occured while browsing sever")
          console.log(err);
        });

      if (result === true && refresh)
        refresh(location.pathname);
    }
    
    setSelName('');
  }

  return (
    <div className="list-presenter">
      {isLoading ? 
        <WaitingComponent /> : 
        <>
          {location.pathname !== '/' && 
            <div className='go-upper' onClick={e => {
              const path = location.pathname.split('/').slice(0, -1).join('/');
              navigate(path);
              if (refresh) refresh(path);
            }}>
              ... go back
            </div>
          }
          {list.map((entry: FileInfo, idx: number) => (
            <ListEntry data={entry} refresh={refresh} showConfirm={showConfirmDlg} key={idx}/>
          ))}
          <ConfirmModal caption={selName} closeModal={closeModal} />
        </>
      }
    </div>
  )
}

export default ListPresenter;
