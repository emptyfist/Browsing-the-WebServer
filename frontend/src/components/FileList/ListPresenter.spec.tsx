import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import ListPresenter from './ListPresenter';

describe('ListPresenter test', function () {
  it('should return only Waiting component whilte loading', function () {
    const { container } = render(
      <MemoryRouter>
        <ListPresenter isLoading={true} list={[]} refresh={() => null} />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it('should return several entries after loading', function () {
    const data = [
      {
        Type: 'Directory',
        Name: 'First Directory',
        Size: 0
      },
      {
        Type: 'Directory',
        Name: 'Second Directory',
        Size: 0
      },
      {
        Type: 'File',
        Name: 'First File',
        Size: 102305
      },
      {
        Type: 'File',
        Name: 'First Directory',
        Size: 1025
      }
    ]
    const { container } = render(
      <MemoryRouter>
        <ListPresenter isLoading={true} list={data} refresh={() => null} />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
