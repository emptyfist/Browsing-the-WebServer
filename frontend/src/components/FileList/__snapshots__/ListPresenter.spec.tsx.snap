// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPresenter test should return only Waiting component whilte loading 1`] = `
<div>
  <div
    class="list-presenter"
  >
    <div
      class="waiting-wrapper"
    >
      <svg
        fill="white"
        height="48"
        viewBox="0 0 24 24"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="10"
          rx="1"
          width="10"
          x="1"
          y="1"
        >
          <animate
            attributeName="x"
            begin="0;l.end"
            dur="0.2s"
            fill="freeze"
            id="a"
            values="1;13"
          />
          <animate
            attributeName="y"
            begin="c.end"
            dur="0.2s"
            fill="freeze"
            id="d"
            values="1;13"
          />
          <animate
            attributeName="x"
            begin="f.end"
            dur="0.2s"
            fill="freeze"
            id="g"
            values="13;1"
          />
          <animate
            attributeName="y"
            begin="i.end"
            dur="0.2s"
            fill="freeze"
            id="j"
            values="13;1"
          />
        </rect>
        <rect
          height="10"
          rx="1"
          width="10"
          x="1"
          y="13"
        >
          <animate
            attributeName="y"
            begin="a.end"
            dur="0.2s"
            fill="freeze"
            id="b"
            values="13;1"
          />
          <animate
            attributeName="x"
            begin="d.end"
            dur="0.2s"
            fill="freeze"
            id="e"
            values="1;13"
          />
          <animate
            attributeName="y"
            begin="g.end"
            dur="0.2s"
            fill="freeze"
            id="h"
            values="1;13"
          />
          <animate
            attributeName="x"
            begin="j.end"
            dur="0.2s"
            fill="freeze"
            id="k"
            values="13;1"
          />
        </rect>
        <rect
          height="10"
          rx="1"
          width="10"
          x="13"
          y="13"
        >
          <animate
            attributeName="x"
            begin="b.end"
            dur="0.2s"
            fill="freeze"
            id="c"
            values="13;1"
          />
          <animate
            attributeName="y"
            begin="e.end"
            dur="0.2s"
            fill="freeze"
            id="f"
            values="13;1"
          />
          <animate
            attributeName="x"
            begin="h.end"
            dur="0.2s"
            fill="freeze"
            id="i"
            values="1;13"
          />
          <animate
            attributeName="y"
            begin="k.end"
            dur="0.2s"
            fill="freeze"
            id="l"
            values="1;13"
          />
        </rect>
      </svg>
    </div>
  </div>
</div>
`;

exports[`ListPresenter test should return several entries after loading 1`] = `
<div>
  <div
    class="list-presenter"
  >
    <div
      class="waiting-wrapper"
    >
      <svg
        fill="white"
        height="48"
        viewBox="0 0 24 24"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="10"
          rx="1"
          width="10"
          x="1"
          y="1"
        >
          <animate
            attributeName="x"
            begin="0;l.end"
            dur="0.2s"
            fill="freeze"
            id="a"
            values="1;13"
          />
          <animate
            attributeName="y"
            begin="c.end"
            dur="0.2s"
            fill="freeze"
            id="d"
            values="1;13"
          />
          <animate
            attributeName="x"
            begin="f.end"
            dur="0.2s"
            fill="freeze"
            id="g"
            values="13;1"
          />
          <animate
            attributeName="y"
            begin="i.end"
            dur="0.2s"
            fill="freeze"
            id="j"
            values="13;1"
          />
        </rect>
        <rect
          height="10"
          rx="1"
          width="10"
          x="1"
          y="13"
        >
          <animate
            attributeName="y"
            begin="a.end"
            dur="0.2s"
            fill="freeze"
            id="b"
            values="13;1"
          />
          <animate
            attributeName="x"
            begin="d.end"
            dur="0.2s"
            fill="freeze"
            id="e"
            values="1;13"
          />
          <animate
            attributeName="y"
            begin="g.end"
            dur="0.2s"
            fill="freeze"
            id="h"
            values="1;13"
          />
          <animate
            attributeName="x"
            begin="j.end"
            dur="0.2s"
            fill="freeze"
            id="k"
            values="13;1"
          />
        </rect>
        <rect
          height="10"
          rx="1"
          width="10"
          x="13"
          y="13"
        >
          <animate
            attributeName="x"
            begin="b.end"
            dur="0.2s"
            fill="freeze"
            id="c"
            values="13;1"
          />
          <animate
            attributeName="y"
            begin="e.end"
            dur="0.2s"
            fill="freeze"
            id="f"
            values="13;1"
          />
          <animate
            attributeName="x"
            begin="h.end"
            dur="0.2s"
            fill="freeze"
            id="i"
            values="1;13"
          />
          <animate
            attributeName="y"
            begin="k.end"
            dur="0.2s"
            fill="freeze"
            id="l"
            values="1;13"
          />
        </rect>
      </svg>
    </div>
  </div>
</div>
`;
